solver/src/single_order_solver.rs

```diff
@@ -0,0 +1,80 @@
+use anyhow::Result;
+use futures::future;
+use rand::prelude::SliceRandom;
+
+use crate::{
+    liquidity::{LimitOrder, Liquidity},
+    settlement::Settlement,
+    solver::Solver,
+};
+
+#[async_trait::async_trait]
+/// Implementations of this trait know how to settle a single limit order (not taking advantage of batching multiple orders together)
+pub trait SingleOrderSolving {
+    /// Return a settlement for the given limit order (if possible)
+    async fn settle_order(&self, order: LimitOrder) -> Result<Settlement>;
+
+    fn name(&self) -> &'static str;
+}
+
+/// Maximum number of sell orders to consider for settlements.
+///
+/// This is mostly out of concern to avoid rate limiting and because
+/// requests may a non-trivial amount of time.
+const MAX_SETTLEMENTS: usize = 5;
+
+pub struct SingleOrderSolver {
+    inner: Box<dyn SingleOrderSolving + Sync + Send>,
+}
+impl<T: SingleOrderSolving + Sync + Send + 'static> From<T> for SingleOrderSolver {
+    fn from(inner: T) -> Self {
+        Self {
+            inner: Box::new(inner),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Solver for SingleOrderSolver {
+    async fn solve(&self, liquidity: Vec<Liquidity>, _gas_price: f64) -> Result<Vec<Settlement>> {
+        let mut orders = liquidity
+            .into_iter()
+            .filter_map(|liquidity| match liquidity {
+                Liquidity::Limit(order) => Some(order),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
```

Comment:
I know this was the point of the single order solver to only operate on actual limit orders. However, I am now struck with the question "how can a single order be matched without any other liquidity sources to consider?" Is this single order solver's purpose entirely for matching orders directly with the settlement contract's reserves? Or is this intended to be for COWs? If the answer is the latter, I don't see how there can be a COW with only a single order.

Reply:
I think it would make sense to pass the remaining liquidity as an additional argument to the SingleOrderSolver. However, since it's not used now (we are just invoking solvers that have knowledge of off-chain liquidity themselves) I didn't do it just yet.